src/library/Query.js

```diff
@@ -33,11 +53,42 @@ function query(httpMethod, deployment, apiKey, qry, payload) {
 }
 
 function limitQuery(httpMethod, deployment, apiKey, queryPath, limit, offset, payload, address) {
-  // validate and normalize limit and offset
-  const queryOptions = buildQueryOptions(limit, offset, address);
+  const separationLimit = 30;
+  const limitChecked = checkLimit(limit);
+  const offsetChecked = checkOffset(offset);
+  const results = {
+    status: null,
+    message: null,
+  };
+  // MBQUERY, MBCUSTOMQUERY has rows in result but MBEVENTS has not
+  const hasRows = !/^events$/.test(queryPath);
+  const rows = [];
+  let queryRows = [];
+
+  let offsetNext = offsetChecked;
+  do {
+    // validate and normalize limit and offset
+    const queryOptions = buildQueryOptions(separationLimit, offsetNext, address);
+    offsetNext += separationLimit;
+
+    // query
+    const queryResult = query(httpMethod, deployment, apiKey, queryPath + queryOptions, payload);
+    queryRows = hasRows ? queryResult.result.rows : queryResult.result;
+    const overflowedLength = (rows.length + queryRows.length) - limitChecked;
+    if (overflowedLength > 0) {
+      // either add the entire array of "queryRows" to "rows", or only up to the total limit
+      // works because if "end" is greater than the length of the array, it only extracts
+      // up to the end of the array
+      rows.push(...queryRows.slice(0, queryRows.length - overflowedLength));
+    } else {
+      rows.push(...queryRows);
+    }
+
+    results.status = queryResult.status;
+    results.message = queryResult.message;
+  } while (limitChecked !== separationLimit && queryRows.length > 0 && rows.length < limitChecked);
```

Comment:
I realized that in some cases the loop is still making more queries than required, and then also realized we can simplify the logic if we never query for more than we want. I.e., if `limit == 45`, then the loop should run twice, querying for 30 results the first time (`separationLimit`) and 15 results the second time (`limitChecked - rows.length`). That also removes the need for a slice operation.

Unfortunately the Github "suggestion" feature doesn't do a great diff job, and there are only a few lines changed in the below. Please have a look and let me know what you think!

```suggestion
  const separationLimit = 30;
  const limitChecked = checkLimit(limit);
  const offsetChecked = checkOffset(offset);
  const results = {
    status: null,
    message: null,
  };
  // MBQUERY, MBCUSTOMQUERY has rows in result but MBEVENTS has not
  const hasRows = !/^events$/.test(queryPath);
  const rows = [];
  let queryRows = [];

  let offsetNext = offsetChecked;
  do {
    // compute the number of rows to query for
    limitNext = Math.min(limitChecked - rows.length, separationLimit);

    // validate and normalize limit and offset
    const queryOptions = buildQueryOptions(limitNext, offsetNext, address);

    // query
    const queryResult = query(httpMethod, deployment, apiKey, queryPath + queryOptions, payload);
    queryRows = hasRows ? queryResult.result.rows : queryResult.result;
    rows.push(...queryRows);

    results.status = queryResult.status;
    results.message = queryResult.message;

    // compute the offset and limit for the next query
    offsetNext += separationLimit;
  } while (rows.length < limitChecked && queryRows.length == limitNext);
```

Reply:
Thanks for suggesting nice cleanup!